File: backend/app/models/product.py
from decimal import Decimal
from typing import Optional

from sqlalchemy import Boolean, ForeignKey, Index, Integer, Numeric, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin


class ProductCategory(Base, TimestampMixin):
    __tablename__ = "product_categories"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(100), unique=True, nullable=False, index=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    display_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    products: Mapped[list["Product"]] = relationship(back_populates="category")


class Product(Base, TimestampMixin):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    category_id: Mapped[int] = mapped_column(
        ForeignKey("product_categories.id", ondelete="RESTRICT"), nullable=False, index=True
    )

    name: Mapped[str] = mapped_column(String(150), nullable=False, index=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    is_returnable: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    hsn_code: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    unit_of_measure: Mapped[str] = mapped_column(String(20), default="Pcs", nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False, index=True)

    category: Mapped[ProductCategory] = relationship(back_populates="products")
    variants: Mapped[list["ProductVariant"]] = relationship(
        back_populates="product", cascade="all, delete-orphan"
    )

    __table_args__ = (
        Index("ix_products_category_active", "category_id", "is_active"),
    )


class ProductVariant(Base, TimestampMixin):
    __tablename__ = "product_variants"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    product_id: Mapped[int] = mapped_column(
        ForeignKey("products.id", ondelete="CASCADE"), nullable=False, index=True
    )

    name: Mapped[str] = mapped_column(String(150), nullable=False, index=True)
    sku_code: Mapped[Optional[str]] = mapped_column(String(50), unique=True, nullable=True, index=True)

    unit_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    cost_price: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    deposit_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), default=0, nullable=False)
    gst_rate: Mapped[Decimal] = mapped_column(Numeric(5, 2), default=0, nullable=False)

    stock_quantity: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)

    product: Mapped[Product] = relationship(back_populates="variants")

    __table_args__ = (
        Index("ix_variants_product_active", "product_id", "is_active"),
    )

File: app/lib/data/models/product.dart
double _d(dynamic v) {
  if (v == null) return 0.0;
  if (v is num) return v.toDouble();
  if (v is String) return double.tryParse(v) ?? 0.0;
  return 0.0;
}

int _i(dynamic v) {
  if (v == null) return 0;
  if (v is int) return v;
  if (v is num) return v.toInt();
  if (v is String) return int.tryParse(v) ?? 0;
  return 0;
}

class ProductCategory {
  final int id;
  final String name;
  final bool isActive;
  ProductCategory({required this.id, required this.name, required this.isActive});
  factory ProductCategory.fromJson(Map<String, dynamic> j) => ProductCategory(
        id: j['id'] as int,
        name: j['name'] as String,
        isActive: j['is_active'] as bool? ?? true,
      );
}

class ProductVariant {
  final int id;
  final int productId;
  final String name;
  final String? skuCode;
  final double unitPrice;
  final double depositAmount;
  final double gstRate;
  final int stockQuantity;
  final bool isActive;
  final String? productName;
  final String? categoryName;
  final bool? isReturnable;
  final String? hsnCode;

  ProductVariant({
    required this.id,
    required this.productId,
    required this.name,
    this.skuCode,
    required this.unitPrice,
    required this.depositAmount,
    required this.gstRate,
    required this.stockQuantity,
    required this.isActive,
    this.productName,
    this.categoryName,
    this.isReturnable,
    this.hsnCode,
  });

  factory ProductVariant.fromJson(Map<String, dynamic> j) => ProductVariant(
        id: j['id'] as int,
        productId: _i(j['product_id']),
        name: j['name'] as String,
        skuCode: j['sku_code'] as String?,
        unitPrice: _d(j['unit_price']),
        depositAmount: _d(j['deposit_amount']),
        gstRate: _d(j['gst_rate']),
        stockQuantity: _i(j['stock_quantity']),
        isActive: j['is_active'] as bool? ?? true,
        productName: j['product_name'] as String?,
        categoryName: j['category_name'] as String?,
        isReturnable: j['is_returnable'] as bool?,
        hsnCode: j['hsn_code'] as String?,
      );

  String get displayName {
    final p = productName ?? '';
    return p.isEmpty ? name : '$p — $name';
  }
}

class Product {
  final int id;
  final int categoryId;
  final String name;
  final bool isReturnable;
  final String? hsnCode;
  final String? unitOfMeasure;
  final bool isActive;
  final List<ProductVariant> variants;

  Product({
    required this.id,
    required this.categoryId,
    required this.name,
    required this.isReturnable,
    this.hsnCode,
    this.unitOfMeasure,
    required this.isActive,
    required this.variants,
  });

  factory Product.fromJson(Map<String, dynamic> j) => Product(
        id: j['id'] as int,
        categoryId: _i(j['category_id']),
        name: j['name'] as String,
        isReturnable: j['is_returnable'] as bool? ?? false,
        hsnCode: j['hsn_code'] as String?,
        unitOfMeasure: j['unit_of_measure'] as String?,
        isActive: j['is_active'] as bool? ?? true,
        variants: (j['variants'] as List? ?? [])
            .map((v) => ProductVariant.fromJson(v as Map<String, dynamic>))
            .toList(),
      );
}
